docs/TypeScript.md

```diff
@@ -1,291 +1,463 @@
-          type: 'string'
+    }, async (request, reply) => {
+      const customerHeader = request.headers['H-Custom']
+      // do something with request data
+      return `logged in!`
+    }) 
+    ```
+7. Build and run and query with the `username` query string option set to anything other than `admin`. The API should now return a HTTP 500 error `{"statusCode":500,"error":"Internal Server Error","message":"Must be admin"}`
+
+🎉 Good work, now you can define interfaces for each route and have strictly typed request and reply instances. Other parts of the Fastify type system rely on generic properties. Make sure to reference the detailed type system documentation below to learn more about what is available.
+
+### JSON Schema
+
+In the last example we used interfaces to define the types for the request querystring and headers. Many users will already be using JSON Schemas to define these properties, and luckily there is a way to transform existing JSON Schemas into TypeScript interfaces!
+
+1. If you did not complete the 'Getting Started' example, go back and follow steps 1-4 first.
+2. Install the `json-schema-to-typescript` module:
+    ```
+    npm i -D json-schema-to-typescript
+    ```
+3. Create a new folder called `schemas` and add two files `params.json` and `querystring.json`. Copy and paste the following schema definitions into the respective files:
+    ```json
+    {
+      "title": "Params Schema",
+      "type": "object",
+      "properties": {
+        "fuzz": { "type": "string" },
+        "buzz": { "type": "integer" }
+      },
+      "additionalProperties": false,
+      "required": ["buzz"]
+    }
+    ```
+    ```json
+    {
+      "title": "Querystring Schema",
+      "type": "object",
+      "properties": {
+        "foo": { "type": "string" },
+        "bar": { "type": "integer" }
+      },
+      "additionalProperties": false,
+      "required": ["foo"]
+    }
+    ```
+4. Create a new folder `types`, leave it empty. Create a new file in the root of your project called `compileSchemaTypes.ts` and copy/paste the following code:
+    ```typescript
+    import path from 'path'
+    import _fs from 'fs'
+    const fs = _fs.promises
+    import { compileFromFile } from 'json-schema-to-typescript'
+
+    const fileOutputPath = (file: string) => {
+      return path.join(__dirname, '/types', `${path.basename(file, '.json')}.d.ts`)
+    }
+
+    const compileSchemas = async () => {
+      try {
+        const files = await fs.readdir(path.join(__dirname, '/schemas'))
+        for (var file of files) {
+          if (path.extname(file) === '.json') {
+            const ts = await compileFromFile(path.join(__dirname, '/schemas', file))
+            await fs.writeFile(fileOutputPath(file), ts)
+          }
         }
+      } catch (error) {
+        console.error(error)
       }
-    },
-    headers: {
-      type: 'object',
-      properties: {
-        a: {
-          type: 'string'
+    }
+
+    compileSchemas()
```

Comment:
I think we should provide this as a npm module instead.

Reply:
https://github.com/Ethan-Arrowood/compile-schemas-to-typescript

Happy to move this to Fastify org or rename it to whatever.

Reply:
I’d move it to the org and add a README

Reply:
Okay readme added. I'll move it to the org now too